### PR TITLE
[6670] Extend Upload dialog to support different endpoints

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/6670

### CODE REVIEW

File: ui/app/src/components/WidgetsAccordion/WidgetsAccordion.tsx [hunk 1]
```diff
@@ -39,18 +43,30 @@ export interface WidgetsAccordionProps {
...
-  const { title, icon, sxs } = props;
-  const [open, setOpen] = useState(false);
+  const { title, id = `WidgetsAccordion:${props.title.replace(/\s/g, '')}`, icon, sxs } = props;
+  const { uuid } = useActiveSite();
+  const user = useActiveUser();
+  const [open, setOpen] = useState(() => getStoredWidgetsAccordion(uuid, user.username, id)?.open ?? false);
   const theme = useTheme();
   const expandedClass = accordionClasses.expanded;
   const contentClass = accordionSummaryClasses.content;
+  // TODO: Incorporate widgets accordion preference removal in user settings
```
Comment (rart) [inline ui/app/src/components/WidgetsAccordion/WidgetsAccordion.tsx]: Fix

File: ui/app/src/components/UploadDialog/util.ts [hunk 2]
```diff
@@ -16,24 +16,46 @@
...
+  method?: XHRUploadOptions['method'];
+  headers?: XHRUploadOptions['headers'];
+  meta?: Record<string, unknown>;
+  allowedMetaFields?: XHRUploadOptions['allowedMetaFields'];
+  useFormData?: boolean;
+  fieldName?: string;
+  showRemoveButtonAfterComplete?: boolean;
+  successfulUploadButton?: UppyDashboardProps['options']['successfulUploadButton'];
+  // TODO: autoProceed doesn't work with at the Dashboard plugin level
+  // autoProceed?: boolean;
```
Comment (rart) [inline ui/app/src/components/UploadDialog/util.ts]: Fix

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 3]
```diff
@@ -96,6 +96,21 @@ function PanelTopBar(props) {
...
+          aria-label={props.i18n('cancelPending')}
+          title={props.i18n('cancelPending')}
+          disabled={props.isAllComplete && props.state.totalProgress === 100}
+          onClick={() => {
+            props.validateFilesPolicy(Object.values(props.files));
+          }}
+        >
+          {props.i18n('uploadFiles')}
+        </button>
+      )}
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]: Let's add this button on the main actions area (the right side): Cancel ------- Add | Proceed | Clear

File: ui/uppy/src/plugins/Dashboard.jsx [hunk 4]
```diff
@@ -284,7 +284,12 @@ export class Dashboard extends UppyDashboard {
     if (this.opts.autoOpenFileEditor) {
       this.uppy.on('files-added', this.#openFileEditorWhenFilesAdded);
     } else {
-      this.uppy.on('files-added', this.validateFilesPolicy);
+      if (this.opts.autoProceed) {
+        this.uppy.on('files-added', this.validateFilesPolicy);
+      } else {
+        this.uppy.on('files-added', this.#generateLargeThumbnailIfSingleFile);
+        this.uppy.on('file-removed', this.#generateLargeThumbnailIfSingleFile);
```
Comment (rart) [inline ui/uppy/src/plugins/Dashboard.jsx]: What's this new file-removed handler for? Why wasn't it needed before?

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/uppy/src/plugins/Dashboard.jsx (hunk 4 above)
Comment (jvega190) [reply ui/uppy/src/plugins/Dashboard.jsx]: This is not needed (was used for something else that was not final), removing

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 5]
```diff
@@ -95,7 +95,6 @@ function PanelTopBar(props) {
           style={{ width: `${props.state.totalProgress}%` }}
         />
       </div>
-
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]: Avoid unnecessary change history. Restore line.

File: ui/app/src/components/UppyDashboard/UppyDashboard.tsx [hunk 6]
```diff
@@ -381,6 +374,9 @@ const translations = defineMessages({
   },
   projectPoliciesNoComply: {
     defaultMessage: 'File "{fileName}" doesn\'t comply with project policies: {detail}'
+  },
+  uploadFiles: {
+    defaultMessage: 'Upload file(s)'
```
Comment (rart) [inline ui/app/src/components/UppyDashboard/UppyDashboard.tsx]:
```suggestion
    defaultMessage: 'Start Uploads'
```

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 7]
```diff
@@ -143,6 +142,20 @@ function PanelTopBar(props) {
             {props.i18n('addMore')}
           </button>
         )}
+        {!props.autoProceed && (
+          <button
+            className="uppy-dashboard-button-base uppy-dashboard-text-button"
+            type="button"
+            aria-label={props.i18n('uploadFiles')}
+            title={props.i18n('uploadFiles')}
+            disabled={props.isAllComplete && props.state.totalProgress === 100}
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]: This seems to disable the button until all are complete (isn't `isAllComplete && totalProgress == 100` redundant?); make sure the button is disabled whilst there are not uploads pending proceed.

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 8]
```diff
@@ -143,6 +142,20 @@ function PanelTopBar(props) {
             {props.i18n('addMore')}
           </button>
         )}
+        {!props.autoProceed && (
+          <button
+            className="uppy-dashboard-button-base uppy-dashboard-text-button"
+            type="button"
+            aria-label={props.i18n('uploadFiles')}
+            title={props.i18n('uploadFiles')}
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]:
```suggestion
```

File: ui/app/src/components/UppyDashboard/UppyDashboard.tsx [hunk 9]
```diff
@@ -381,6 +374,9 @@ const translations = defineMessages({
   },
   projectPoliciesNoComply: {
     defaultMessage: 'File "{fileName}" doesn\'t comply with project policies: {detail}'
+  },
+  uploadFiles: {
```
Comment (rart) [inline ui/app/src/components/UppyDashboard/UppyDashboard.tsx]:
```suggestion
  proceed: {
```

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 10]
```diff
@@ -143,6 +142,20 @@ function PanelTopBar(props) {
...
+            disabled={props.isAllComplete && props.state.totalProgress === 100}
+            onClick={() => {
+              props.validateFilesPolicy(Object.values(props.files));
+            }}
+          >
+            {props.i18n('uploadFiles')}
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]:
```suggestion
            {props.i18n('proceed')}
```

**Review by rart — CHANGES_REQUESTED** [review]
It would be nice to have a icon button per upload to proceed on an individual basis as well.

File: ui/uppy/src/components/PickerPanelTopBar.jsx [hunk 11]
```diff
@@ -76,7 +76,30 @@ function UploadStatus({
...
+  const uploadingState = getUploadingState(isAllErrored, isAllComplete, isAllPaused, files);
+  const invalidFilesCount = Object.values(invalidFiles).filter((invalid) => invalid).length;
+  const newFilesCount = newFiles.length;
+  const disableProceed =
+    // if there are active uploads
+    uploadingState === 'uploading' ||
+    // there are no new files to upload (all files already uploaded)
+    !newFilesCount ||
+    // if there are invalid files (due to site policies) waiting to be uploaded
+    (invalidFilesCount && uploadingState === 'waiting');
```
Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]: - This looks like if files are added whilst other files are being uploaded, the proceed will be disabled?
- If newFilesCount is a number, let's express that condition as dealing with numbers (i.e. newFilesCount === 0)

Comment (rart) [inline ui/uppy/src/components/PickerPanelTopBar.jsx]: - This looks like if files are added whilst other files are being uploaded, the proceed will be disabled?
- If newFilesCount is a number, let's express that condition as dealing with numbers (i.e. newFilesCount === 0)

**Review by rart — CHANGES_REQUESTED** [review]
- What happens if files are added whilst others are being uploaded?
- Can/did we add the individual proceed button?

File: ui/uppy/src/components/PickerPanelTopBar.jsx (hunk 11 above)
Comment (jvega190) [reply ui/uppy/src/components/PickerPanelTopBar.jsx]: this comment is repeated, I'll resolve this one and keep the other one until the details are actually resolved

Comment (jvega190) [reply ui/uppy/src/components/PickerPanelTopBar.jsx]: Updated, it'll go like this:
With `autoproceed` false, if you add files and hit the `Start Uploads` button, it'll start uploading the added files, and `Start Uploads` will be disabled.
If while uploading, more files are added, then the `Start Uploads` button will be re-enabled, and if the button is clicked, it'll start uploading only the newly added files (the previous uploads continue).

disableProceed validation became simpler because now `newFiles` doesn't include the files pending due to policies validation (already started to upload), so by only validating the newFiles count is enough.

Also, I missed the individual proceed comment, it is now added, and the proceed button it'll only show at the beginning (when autoProceed is false), and the batch proceed button works as expected (not disabled unless there are no pending uploads)

File: ui/uppy/src/components/FileItem/Buttons/index.jsx [hunk 12]
```diff
@@ -55,7 +70,13 @@ function RetryButton({ i18n, onClick }) {
 function ValidateIcon({ i18n }) {
   return (
     <i className="uppy-dashboard-item-validating" title={i18n('validating')}>
-      <svg className="uppy-dashboard-svg-icon" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
+      <svg
+        className="uppy-dashboard-svg-icon"
+        focusable="false"
+        viewBox="0 0 24 24"
+        aria-hidden="true"
+        style={{ color: '#757575' }}
```
Comment (rart) [inline ui/uppy/src/components/FileItem/Buttons/index.jsx]: Is this theme-friendly? (dark/light)
What's the need to change this now?

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/uppy/src/components/FileItem/Buttons/index.jsx (hunk 12 above)
Comment (jvega190) [reply ui/uppy/src/components/FileItem/Buttons/index.jsx]: Yeah it maches our dark theme. There was this visual issue (I guess it was never catched) when there were uploaded files, if you clicked 'Add more' the view didn't have dark mode styles.

Before fix
Screenshot from 2024-05-06 12-03-14
After fix
Screenshot from 2024-05-06 12-56-10

Comment (rart) [reply ui/uppy/src/components/FileItem/Buttons/index.jsx]: And light?

Comment (rart) [reply ui/uppy/src/components/FileItem/Buttons/index.jsx]: Ping to review

Comment (rart) [reply ui/uppy/src/components/FileItem/Buttons/index.jsx]: Where are we at with this? The colour is still hardcoded to the **icon**.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/uppy/src/components/FileItem/Buttons/index.jsx (hunk 12 above)
Comment (jvega190) [reply ui/uppy/src/components/FileItem/Buttons/index.jsx]: Updated the Icon so it uses styles from proper class, removed hardcoded styles

File: ui/uppy/src/components/FileItem/Buttons/index.jsx [hunk 13]
```diff
@@ -79,18 +94,64 @@ function AcceptSuggestedNameIcon({ i18n, onClick }) {
...
+    i18n,
+    removeFile,
+    error,
+    hideRetryButton,
+    retryUpload,
+    validateAndRetry,
+    successfulUploadButton,
+    validateFilesPolicy
+  } = props;
+  console.log('props', props);
```
Comment (rart) [inline ui/uppy/src/components/FileItem/Buttons/index.jsx]: x

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — APPROVED** [review]
